File: 2021/day03/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

std::vector<std::string> parseInput() {
    std::vector<std::string> numbers;
    std::ifstream file("../input.txt");
    std::string line;

    while (std::getline(file, line)) {
        if (!line.empty()) {
            numbers.push_back(line);
        }
    }

    return numbers;
}

int part1(const std::vector<std::string>& numbers) {
    int numBits = numbers[0].length();
    int gamma = 0;

    for (int pos = 0; pos < numBits; pos++) {
        int ones = 0;
        for (const auto& n : numbers) {
            if (n[pos] == '1') {
                ones++;
            }
        }
        int zeros = numbers.size() - ones;

        if (ones >= zeros) {
            gamma |= (1 << (numBits - 1 - pos));
        }
    }

    // epsilon is bitwise NOT of gamma (within numBits)
    int epsilon = gamma ^ ((1 << numBits) - 1);

    return gamma * epsilon;
}

int findRating(const std::vector<std::string>& numbers, bool useMostCommon) {
    int numBits = numbers[0].length();
    std::vector<std::string> candidates = numbers;

    for (int pos = 0; pos < numBits; pos++) {
        if (candidates.size() == 1) {
            break;
        }

        int ones = 0;
        for (const auto& n : candidates) {
            if (n[pos] == '1') {
                ones++;
            }
        }
        int zeros = candidates.size() - ones;

        char target;
        if (useMostCommon) {
            target = (ones >= zeros) ? '1' : '0';
        } else {
            target = (zeros <= ones) ? '0' : '1';
        }

        std::vector<std::string> filtered;
        for (const auto& n : candidates) {
            if (n[pos] == target) {
                filtered.push_back(n);
            }
        }
        candidates = filtered;
    }

    // Convert binary string to int
    int result = 0;
    for (char c : candidates[0]) {
        result = (result << 1) | (c - '0');
    }
    return result;
}

int part2(const std::vector<std::string>& numbers) {
    int oxygen = findRating(numbers, true);
    int co2 = findRating(numbers, false);
    return oxygen * co2;
}

int main() {
    std::vector<std::string> numbers = parseInput();

    std::cout << "Part 1: " << part1(numbers) << std::endl;
    std::cout << "Part 2: " << part2(numbers) << std::endl;

    return 0;
}

File: 2021/day03/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
)

func parseInput() ([]string, error) {
	exePath, err := os.Executable()
	if err != nil {
		return nil, err
	}
	inputPath := filepath.Join(filepath.Dir(exePath), "..", "input.txt")

	file, err := os.Open(inputPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var numbers []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line != "" {
			numbers = append(numbers, line)
		}
	}
	return numbers, scanner.Err()
}

func part1(numbers []string) int {
	numBits := len(numbers[0])
	gamma := 0

	for pos := 0; pos < numBits; pos++ {
		ones := 0
		for _, n := range numbers {
			if n[pos] == '1' {
				ones++
			}
		}
		zeros := len(numbers) - ones

		if ones >= zeros {
			gamma |= 1 << (numBits - 1 - pos)
		}
	}

	// epsilon is bitwise NOT of gamma (within numBits)
	epsilon := gamma ^ ((1 << numBits) - 1)

	return gamma * epsilon
}

func binaryToInt(s string) int {
	result := 0
	for _, c := range s {
		result <<= 1
		if c == '1' {
			result |= 1
		}
	}
	return result
}

func findRating(numbers []string, useMostCommon bool) int {
	numBits := len(numbers[0])
	candidates := make([]string, len(numbers))
	copy(candidates, numbers)

	for pos := 0; pos < numBits; pos++ {
		if len(candidates) == 1 {
			break
		}

		ones := 0
		for _, n := range candidates {
			if n[pos] == '1' {
				ones++
			}
		}
		zeros := len(candidates) - ones

		var target byte
		if useMostCommon {
			if ones >= zeros {
				target = '1'
			} else {
				target = '0'
			}
		} else {
			if zeros <= ones {
				target = '0'
			} else {
				target = '1'
			}
		}

		var filtered []string
		for _, n := range candidates {
			if n[pos] == target {
				filtered = append(filtered, n)
			}
		}
		candidates = filtered
	}

	return binaryToInt(candidates[0])
}

func part2(numbers []string) int {
	oxygen := findRating(numbers, true)
	co2 := findRating(numbers, false)
	return oxygen * co2
}

func main() {
	numbers, err := parseInput()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Part 1: %d\n", part1(numbers))
	fmt.Printf("Part 2: %d\n", part2(numbers))
}
